WIP Reactions backend: removing reactions

arpith opened the pull request:
This PR adds support for removing reactions via `DELETE` requests to the `/reactions` endpoint with parameters `emoji_name` and `message_id`.

The reaction is deleted from the db and a reaction event is sent out with `op` set to `remove`.

Tests are added to check that you can't remove a reaction that does not exist and to check that the event payload and users are correct for a delete request.

smarx commented:
*Automated message from Dropbox CLA bot*

@arpith, it looks like you've already signed the Dropbox CLA. Thanks!

arpith commented:
@timabbott, @kracekumar could you take a look at this?

timabbott commented:
This looks pretty reasonable; my only concern is that I'm not sure that use of DELETE without a specific resource is proper REST API style.  We might want to redo the endpoints for reactions to be of the form: `PUT|DELETE /messages/1234/emoji_reactions/banana` to follow proper REST style.

What do you think @arpith @kracekumar?

timabbott commented:
I'm fine with doing the update to the endpoints as an additional commit on top of this, since we need to change the existing ADD endpoint anyway, so I'm going to ahead and merge this.

timabbott commented:
@arpith one thing I notice in the coverage data is that removing a nonexistent reaction doesn't seem to have coverage -- can you look into why?

arpith commented:
Thanks @timabbott :) The removing a nonexistent reaction test is here:
https://github.com/zulip/zulip/blob/1fb9220354ec16feb12e794d1f58bc8b5cd668d0/zerver/tests/test_reactions.py#L144
Will take a look to see why it's not reflected in the coverage data.

kracekumar commented:
I'm late to the party but +1 for `REST` style `DELETE /messages/1234/emoji_reactions/banana`. Many HTTP clients don't t support `DELETE` with the body.

arpith commented:
The coverage drop was testing invalid emoji's when removing reactions - I've added a test in https://github.com/zulip/zulip/pull/2570 and also fixed the endpoints

arpith commented:
Never too late @kracekumar :) fixed in https://github.com/zulip/zulip/pull/2570

kracekumar commented:
Now I think it's better to use `DELETE /messages/1234/emoji_reactions/<reaction_id>` since `reaction_id` is `PK` and  `emoji_name` isn't indexed. This means `reaction id` needs to be added in the API.

timabbott commented:
Also, we can change what the primary key is in the DB model if needed; we should design the API to be best for the programmers coding against it, not for technical limitations, unless they're really insurmountable.

kracekumar commented:
That's very good point!
